rcljava/src/main/java/org/ros2/rcljava/action/ActionServerImpl.java

```diff
@@ -0,0 +1,561 @@
+ * limitations under the License.
+ */
+
+package org.ros2.rcljava.action;
+
+import java.lang.ref.WeakReference;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.ros2.rcljava.RCLJava;
+import org.ros2.rcljava.common.JNIUtils;
+import org.ros2.rcljava.consumers.Consumer;
+import org.ros2.rcljava.interfaces.MessageDefinition;
+import org.ros2.rcljava.interfaces.ActionDefinition;
+import org.ros2.rcljava.interfaces.GoalRequestDefinition;
+import org.ros2.rcljava.interfaces.GoalResponseDefinition;
+import org.ros2.rcljava.node.Node;
+import org.ros2.rcljava.service.RMWRequestId;
+import org.ros2.rcljava.time.Clock;
+import org.ros2.rcljava.Time;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ActionServerImpl<T extends ActionDefinition> implements ActionServer<T> {
+  static class GoalHandleImpl<T extends ActionDefinition>  implements ActionServerGoalHandle {
+    private static final Logger logger = LoggerFactory.getLogger(GoalHandleImpl.class);
+
+    static {
+      try {
+        JNIUtils.loadImplementation(GoalHandleImpl.class);
+      } catch (UnsatisfiedLinkError ule) {
+        logger.error("Native code library failed to load.\n" + ule);
+        System.exit(1);
+      }
+    }
+
+    private long handle;
+    private ActionServer<T> actionServer;
+    private action_msgs.msg.GoalInfo goalInfo;
+    private MessageDefinition goal;
+
+    private static native long nativeAcceptNewGoal(
+      long actionServerHandle,
+      long goalInfoFromJavaConverterHandle,
+      long goalInfoDestructorHandle,
+      MessageDefinition goalInfo);
+    private static native int nativeGetStatus(long goalHandle);
+    private static native void nativeGoalEventExecute(long goalHandle);
+    private static native void nativeGoalEventCancelGoal(long goalHandle);
+    private static native void nativeGoalEventSucceed(long goalHandle);
+    private static native void nativeGoalEventAbort(long goalHandle);
+    private static native void nativeGoalEventCanceled(long goalHandle);
+    private static native void nativeDispose(long handle);
+
+    public GoalHandleImpl(
+      ActionServer<T> actionServer, action_msgs.msg.GoalInfo goalInfo, MessageDefinition goal)
+    {
+      this.actionServer = actionServer;
+      this.goalInfo = goalInfo;
+      this.goal = goal;
+      long goalInfoFromJavaConverterHandle = goalInfo.getFromJavaConverterInstance();
+      long goalInfoDestructorHandle = goalInfo.getDestructorInstance();
+      this.handle = nativeAcceptNewGoal(
+        actionServer.getHandle(),
+        goalInfoFromJavaConverterHandle,
+        goalInfoDestructorHandle,
+        goalInfo);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public synchronized action_msgs.msg.GoalInfo getGoalInfo() {
+      return this.goalInfo;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public synchronized MessageDefinition getGoal() {
```

Comment:
Considering that the `GoalHandleImpl` is not public, maybe it's not important here, but something like this can make writing `goalCallback`/`cancelCallback` easier

Reply:
It seems like a trade-off: we can let the user avoid a potentially unsafe cast at the cost of verbose constructor and method calls. Each additional type parameter bloats the construction of an action server, e.g.

```java
class MyCancelCallback implements CancelCallback<test_msgs.action.Fibonacci, test_msgs.action.Fibonacci.Goal, ...> {
  public CancelResponse handleCancel(ActionServerGoalHandle<test_msgs.action.Fibonacci, test_msgs.action.Fibonacci.Goal, ...> goalHandle) {
    // Now we don't have to cast
    test_msgs.action.Fibonacci.Goal = goalHandle.getGoal();
    return CancelResponse.ACCEPT;
  }
}

...

node.<test_msgs.action.Fibonacci, test_msgs.action.Fibonacci.Goal, ...>createActionServer(
  test_msgs.action.Fibonacci.class, "test_action",
  goalCallback, cancelCallback, acceptedCallback);
```

versus

```java
class MyCancelCallback implements CancelCallback<test_msgs.action.Fibonacci> {
  public CancelResponse handleCancel(ActionServerGoalHandle<test_msgs.action.Fibonacci> goalHandle) {
    // Now we should cast
    test_msgs.action.Fibonacci.Goal = (test_msgs.action.Fibonacci.Goal) goalHandle.getGoal();
    return CancelResponse.ACCEPT;
  }
}

...

node.<test_msgs.action.Fibonacci>createActionServer(
  test_msgs.action.Fibonacci.class, "test_action",
  goalCallback, cancelCallback, acceptedCallback);
```

I'm not sure what best practices for Java would dictate in this situation.